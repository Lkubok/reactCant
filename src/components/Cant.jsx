import React, { Component } from "react";
import {
  updateProfileSlopeValue,
  updateOffsetValue,
  updateNormalCrown,
  updateFullCrown,
  updateLengthOfCant,
  updateDesignSpeed
} from "../actions";

import { connect } from "react-redux";
import CantParams from "./CantParams";
import LeftRightSelect from "./LeftRightSelect";
import * as selectors from "../reducers/selectors";

export class Cant extends Component {
  constructor(props) {
    super(props);
    this.handleOnChange = this.handleOnChange.bind(this);
  }
  handleOnChange(e) {
    console.log("koko");
    console.log(e);
    let { name, value } = e.target;
    value = parseFloat(value);
    if (name === "mainSlope") this.props.changeProfileSlope(value);
    if (name === "laneOffset") this.props.changeOffsetValue(value);
    if (name === "normalCrown") this.props.updateNormalCrown(value);
    if (name === "fullCrown") this.props.updateFullCrown(value);
    if (name === "lengthOfCant") this.props.updateLengthOfCant(value);
    if (name === "designSpeed") this.props.updateDesignSpeed(value);
  }

  handleOnChangeNew = callback => e => {
    let { value } = e.target;
    value = parseFloat(value);
    callback(value);
  };
  /*   handleCalculate() {
    const {
      profileSlopeValue: profile,
      laneOffsetValue: lane,
      normalCrownSlope: normal,
      fullSuperSlope: full,
      lengthOfCant: length
    } = this.props;
    console.log(profile, lane, normal, full, length);

    const addSlope = ((Math.abs(normal) + Math.abs(full)) * lane) / length;
    this.props.updateAdditionalSlope(addSlope); // IT HAPPENS TWICE... WHY  ????
    return addSlope;
  } */
  checkAddSlope = () => {
    const { designSpeed: speed, result, laneOffsetValue: lane } = this.props;
    if (speed >= 100)
      return result <= 0.9 && result > lane * 0.1
        ? "alert-success"
        : "alert-danger";
    if (speed >= 80)
      return result <= 1.0 && result > lane * 0.1
        ? "alert-success"
        : "alert-danger";
    if (speed >= 60)
      return result <= 1.6 && result > lane * 0.1
        ? "alert-success"
        : "alert-danger";
    return result <= 2.0 && result > lane * 0.1
      ? "alert-success"
      : "alert-danger";
  };
  render() {
    return (
      <>
        <form className="mt-5">
          <div className="form-row">
            <div className="form-group col-md-4">
              <label htmlFor="mainSlope">
                Main profile slope in %: (up is +, down is -)
              </label>
              <input
                onChange={this.handleOnChangeNew(this.props.changeProfileSlope)}
                value={this.props.profileSlopeValue}
                type="number"
                step="0.01"
                max="15"
                min="-15"
                className="form-control"
                name="mainSlope"
                id="mainSlope"
              />
            </div>
            <div className="form-group col-md-4">
              <label htmlFor="laneOffset">
                Offset to edge of lane from rotation axis [m]
              </label>
              <input
                onChange={this.handleOnChange}
                value={this.props.laneOffsetValue}
                type="number"
                step="0.1"
                className="form-control"
                name="laneOffset"
                id="laneOffset"
              />
            </div>
            <div className="form-group col-md-4">
              <label htmlFor="designSpeed">Design speed [km/h]</label>
              <input
                onChange={this.handleOnChange}
                value={this.props.designSpeed}
                type="number"
                step="10"
                min="30"
                max="120"
                className="form-control"
                name="designSpeed"
                id="designSpeed"
              />
            </div>
          </div>

          <div className="form-row">
            <div className="form-group col-md-4">
              <label htmlFor="normalCrown">
                Normal Crown Slope [%] (- is down)
              </label>
              <input
                onChange={this.handleOnChange}
                value={this.props.normalCrownSlope}
                type="number"
                step="0.1"
                max="15"
                min="-15"
                className="form-control"
                name="normalCrown"
                id="normalCrown"
              />
            </div>
            <div className="form-group col-md-4">
              <label htmlFor="fullCrown">Full Super Slope [%]</label>
              <input
                onChange={this.handleOnChange}
                value={this.props.fullSuperSlope}
                type="number"
                step="0.1"
                className="form-control"
                name="fullCrown"
                id="fullCrown"
              />
            </div>
            <div className="form-group col-md-4">
              <label htmlFor="lengthOfCant">Length of Cant [m]</label>
              <input
                onChange={this.handleOnChange}
                value={this.props.lengthOfCant}
                type="number"
                step="0.1"
                className="form-control"
                name="lengthOfCant"
                id="lengthOfCant"
              />
            </div>
          </div>

          <div className="form-group">
            <label htmlFor="additionalSlope">
              Additional slope on Edges [%]
            </label>
            <div className={`form-control ${this.checkAddSlope()}`}>
              {this.props.result}
            </div>
          </div>
          <LeftRightSelect />
          <CantParams />
        </form>
      </>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  result: selectors.addEdgeSelector(state, 1),
  fullSuperSlope: selectors.getFullSuperSlope(state),
  lengthOfCant: selectors.getLengthOfCant(state),
  normalCrownSlope: selectors.getNormalCrownSlope(state),
  laneOffsetValue: selectors.getLaneOffsetValue(state),
  profileSlopeValue: selectors.getProfileSlopeValue(state),
  designSpeed: selectors.getDesignSpeedValue(state)
});

const mapDispatchToProps = dispatch => ({
  changeProfileSlope: arg => dispatch(updateProfileSlopeValue(arg)),
  changeOffsetValue: arg => dispatch(updateOffsetValue(arg)),
  updateNormalCrown: arg => dispatch(updateNormalCrown(arg)),
  updateFullCrown: arg => dispatch(updateFullCrown(arg)),
  updateLengthOfCant: arg => dispatch(updateLengthOfCant(arg)),
  updateDesignSpeed: arg => dispatch(updateDesignSpeed(arg))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Cant);
