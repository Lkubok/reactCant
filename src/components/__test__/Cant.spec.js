import { Cant } from "../Cant";
import { shallow } from "enzyme";
import React from "react";

let props;

describe("Cant Component", () => {
  beforeEach(() => {
    props = {
      changeProfileSlope: jest.fn(),
      changeOffsetValue: jest.fn(),
      fullSuperSlope: 123,
      laneOffsetValue: 1234
    };
  });
  it("should mount", () => {
    const wrapper = shallow(<Cant {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
  /*   it("Trigger handle Profile Slope when handling event", () => {
    const wrapper = shallow(<Cant {...props} />);
    const testObj = {
      target: {
        name: "mainSlope",
        value: 2
      }
    };
    wrapper.instance().handleOnChange(testObj);
    expect(props.changeProfileSlope).toHaveBeenCalledTimes(1);
    expect(props.changeProfileSlope).toHaveBeenCalledWith(2);
    expect(props.changeOffsetValue).not.toHaveBeenCalled();
  }); */
  it("Render input LaneOffset value with value from props", () => {
    const wrapper = shallow(<Cant {...props} />);
    expect(wrapper.find("#laneOffset").length).toEqual(1);
    expect(wrapper.find("#laneOffset").props().value).toEqual(1234);
  });
  it("Onchange on input should trigger hadleChange", () => {
    const wrapper = shallow(<Cant {...props} />);
    const fancySpy = jest.fn();
    const event = {
      target: {
        name: "laneOffsetValue",
        value: 2
      }
    };
    wrapper.instance().handleOnChange = fancySpy;
    wrapper.update();
    wrapper.find("#laneOffset").simulate("change", event);
    expect(fancySpy).toHaveBeenCalled();
    // expect(true).toBe(true);
  });
});

// if (name === "mainSlope") this.props.changeProfileSlope(value);
