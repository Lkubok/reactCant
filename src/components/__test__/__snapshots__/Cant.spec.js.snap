// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cant Component should mount 1`] = `
<Fragment>
  <form
    className="mt-5"
  >
    <div
      className="form-row"
    >
      <div
        className="form-group col-md-4"
      >
        <label
          htmlFor="mainSlope"
        >
          Main profile slope in %: (up is +, down is -)
        </label>
        <input
          className="form-control"
          id="mainSlope"
          max="15"
          min="-15"
          name="mainSlope"
          onChange={[Function]}
          step="0.01"
          type="number"
        />
      </div>
      <div
        className="form-group col-md-4"
      >
        <label
          htmlFor="laneOffset"
        >
          Offset to edge of lane from rotation axis [m]
        </label>
        <input
          className="form-control"
          id="laneOffset"
          name="laneOffset"
          onChange={[Function]}
          step="0.1"
          type="number"
          value={1234}
        />
      </div>
      <div
        className="form-group col-md-4"
      >
        <label
          htmlFor="designSpeed"
        >
          Design speed [km/h]
        </label>
        <input
          className="form-control"
          id="designSpeed"
          max="120"
          min="30"
          name="designSpeed"
          onChange={[Function]}
          step="10"
          type="number"
        />
      </div>
    </div>
    <div
      className="form-row"
    >
      <div
        className="form-group col-md-4"
      >
        <label
          htmlFor="normalCrown"
        >
          Normal Crown Slope [%] (- is down)
        </label>
        <input
          className="form-control"
          id="normalCrown"
          max="15"
          min="-15"
          name="normalCrown"
          onChange={[Function]}
          step="0.1"
          type="number"
        />
      </div>
      <div
        className="form-group col-md-4"
      >
        <label
          htmlFor="fullCrown"
        >
          Full Super Slope [%]
        </label>
        <input
          className="form-control"
          id="fullCrown"
          name="fullCrown"
          onChange={[Function]}
          step="0.1"
          type="number"
          value={123}
        />
      </div>
      <div
        className="form-group col-md-4"
      >
        <label
          htmlFor="lengthOfCant"
        >
          Length of Cant [m]
        </label>
        <input
          className="form-control"
          id="lengthOfCant"
          name="lengthOfCant"
          onChange={[Function]}
          step="0.1"
          type="number"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="additionalSlope"
      >
        Additional slope on Edges [%]
      </label>
      <div
        className="form-control alert-danger"
      />
    </div>
    <Connect(LeftRightSelect) />
    <Connect(CantParams) />
  </form>
</Fragment>
`;
